File: shopro-pos-flutter/lib/core/security/dpop_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:cryptography/cryptography.dart';

class DPoPService {
  // We use a persistent key pair for the session.
  // In a real app, this would be generated once and stored in secure storage.
  static KeyPair? _keyPair;
  static Map<String, dynamic>? _jwk;

  static String _generateRandomString(int length) {
    final rand = Random.secure();
    final values = List<int>.generate(length, (i) => rand.nextInt(256));
    return base64Url.encode(values).replaceAll('=', '');
  }

  static Future<void> _ensureKeyPair() async {
    if (_keyPair != null) return;

    final algorithm = Ed25519();
    _keyPair = await algorithm.newKeyPair();
    final publicKey = await _keyPair!.extractPublicKey();
    final bytes = (publicKey as SimplePublicKey).bytes;

    // Construct JWK for Ed25519 (OKP kty)
    _jwk = {
      'kty': 'OKP',
      'crv': 'Ed25519',
      'x': base64Url.encode(bytes).replaceAll('=', ''),
      'use': 'sig',
    };
  }

  /// Generates a DPoP proof JWT for the given request.
  /// [htm] HTTP Method (e.g., POST)
  /// [htu] HTTP URL (Absolute URL without query params)
  static Future<String> generateProof(String htm, String htu) async {
    await _ensureKeyPair();
    
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    
    // Construct Header and Payload manually for precise control
    final header = {
      'typ': 'dpop+jwt',
      'alg': 'EdDSA',
      'jwk': _jwk,
    };
    
    final payload = {
      'jti': _generateRandomString(16),
      'htm': htm.toUpperCase(),
      'htu': htu.split('?')[0], // strip query params per RFC 9449
      'iat': now,
      'exp': now + 120,
    };

    final encodedHeader = base64Url.encode(utf8.encode(jsonEncode(header))).replaceAll('=', '');
    final encodedPayload = base64Url.encode(utf8.encode(jsonEncode(payload))).replaceAll('=', '');
    
    final messageToSign = '$encodedHeader.$encodedPayload';
    
    // Sign using pure cryptography package to avoid dart_jsonwebtoken compatibility issues with Java JJWT
    final algorithm = Ed25519();
    final signature = await algorithm.sign(
      utf8.encode(messageToSign),
      keyPair: _keyPair!,
    );
    
    final encodedSignature = base64Url.encode(signature.bytes).replaceAll('=', '');
    
    return '$messageToSign.$encodedSignature';
  }
}

File: shopro-sabz-web/src/lib/security/dpop-service.ts
import { Base64 } from 'js-base64';

export class DPoPWebService {
  private static cachedKeyPair: CryptoKeyPair | null = null;

  /**
   * Generates or retrieves a persistent, non-extractable RSA-PSS key pair.
   */
  static async getOrCreateKeyPair(): Promise<CryptoKeyPair> {
    if (this.cachedKeyPair) {
      return this.cachedKeyPair;
    }
    const keyPair = await window.crypto.subtle.generateKey(
      {
        name: "RSA-PSS",
        modulusLength: 2048,
        publicExponent: new Uint8Array([1, 0, 1]),
        hash: "SHA-256",
      },
      false, // non-extractable
      ["sign", "verify"]
    );
    this.cachedKeyPair = keyPair;
    return keyPair;
  }

  /**
   * Generates a DPoP proof JWT for the given method and URL.
   */
  static async generateProof(method: string, url: string): Promise<string> {
    const keyPair = await this.getOrCreateKeyPair();
    const publicJWK = await window.crypto.subtle.exportKey("jwk", keyPair.publicKey);

    const header = {
      typ: "dpop+jwt",
      alg: "PS256",
      jwk: {
        kty: publicJWK.kty,
        n: publicJWK.n,
        e: publicJWK.e,
      },
    };

    const payload = {
      jti: crypto.randomUUID(),
      htm: method.toUpperCase(),
      htu: url.split('?')[0], // strip query params
      iat: Math.floor(Date.now() / 1000),
    };

    const encodedHeader = Base64.encodeURI(JSON.stringify(header));
    const encodedPayload = Base64.encodeURI(JSON.stringify(payload));
    const dataToSign = new TextEncoder().encode(`${encodedHeader}.${encodedPayload}`);

    const signature = await window.crypto.subtle.sign(
      {
        name: "RSA-PSS",
        saltLength: 32,
      },
      keyPair.privateKey,
      dataToSign
    );

    const encodedSignature = Base64.fromUint8Array(new Uint8Array(signature), true);
    return `${encodedHeader}.${encodedPayload}.${encodedSignature}`;
  }
}
